ann_benchmarks/datasets.py

```diff
@@ -279,6 +283,50 @@ def sift_hamming(out_fn, fn):
         X_train, X_test = train_test_split(X, test_size=1000)
         write_output(X_train, X_test, out_fn, 'hamming', 'bit')
 
+def kosarak(out_fn):
+    import gzip
+    local_fn = 'kosarak.dat.gz'
+    # only consider sets with at least min_elements many elements
+    min_elements = 20
+    url = 'http://fimi.uantwerpen.be/data/%s' % local_fn
+    download(url, local_fn)
+
+    with gzip.open('kosarak.dat.gz', 'r') as f:
+        content = f.readlines()
+        # preprocess data to find sets with more than 20 elements
+        # keep track of used ids for reenumeration
+        ids = {}
+        next_id = 0
+        cnt = 0
+        for line in content:
+            if len(line.split()) >= min_elements:
+                cnt += 1
+                for x in line.split():
+                    if int(x) not in ids:
+                        ids[int(x)] = next_id
+                        next_id += 1
+
+    X = numpy.zeros((cnt, len(ids)), dtype=numpy.bool)
+    i = 0
+    for line in content:
+        if len(line.split()) >= min_elements:
+            for x in line.split():
+                X[i][ids[int(x)]] = 1
+            i += 1
+
+    X_train, X_test = train_test_split(numpy.array(X), test_size=500)
+    write_output(X_train, X_test, out_fn, 'jaccard', 'bit')
+
+def random_jaccard(out_fn, n=10000, size=50, universe=80):
```

Comment:
can you seed the random here just so it's reproducible?

Reply:
Sure! Added. (There are multipe `random_state=1` and `seed(1)` calls now. Maybe there should be a unique global seed.)